modulemd_tools: add the first pieces of a python library

FrostyX opened the pull request:
Since the beginning of this project (in fact years before that) we
were always talking about creating a python library that would all our
tools use and share code with each other. A library that also others
(such as build-systems or other developers) could use.

I was hesitant to implement it because I was afraid that it will end
up reimplementing the whole `python3-libmodulemd` library (which would
be no good) or that we will create some nice, object-oriented and
high-level API, that would not fit some corner-cases and therefore
would be useless, because other people will have to implement it once
again but differently. I am finaly confident to add the first pieces
of the API.

As I see, it, there are three abstraction layers, that can be
implemented:

1. Functions, that would take modulemd YAMLs as a string, do something
   and return a new modulemd YAML string as their output. These can be
   easily implemented as pure functions with some small exceptions.
2. Functions, that would take and return objects from
   `python3-libmodulemd`
3. Collection of classes for representing modules, packages, etc and
   providing a facade over `python3-libmodule` objects, that would be
   used just internally and user will not have any knowing that they
   are used.

If requested or needed, I am going to implement all three layers,
separated that everyone could use the best approach for their
task. Now I am picking one, that I think will cover the most use-cases
with the least possible complexity, and that is (1).

I believe that this package will allow us to throw away a lot of
modularity-related code from Copr and change it to one-liners.

Lastly, the fact that I spent implementing this simple set of
functions almost three work days, shall be taken as a proof that we
really need such library. Others would have to do the same when
implementing modularity support for other tools / build systems, and
that would be a lot of time wasted.

dmach commented on modulemd_tools/modulemd_tools/yaml.py:
```diff
@@ -0,0 +1,183 @@
+    Determine whether the `yaml` string is a valid modulemd YAML definition
+    """
+    idx = Modulemd.ModuleIndex.new()
+    (ret, _) = idx.update_from_string(yaml, strict=True)
```
The parentheses are not necessary, they even decrease code readability.
I'd use ```ret, _ = ...```

dmach commented on modulemd_tools/modulemd_tools/yaml.py:
```diff
@@ -0,0 +1,183 @@
+    Same as `is_valid` but raises exception if the YAML is not valid.
+    """
+    idx = Modulemd.ModuleIndex.new()
+    (ret, failures) = idx.update_from_string(yaml, strict=True)
```
Also remove the parentheses?

dmach commented on modulemd_tools/modulemd_tools/yaml.py:
```diff
@@ -0,0 +1,183 @@
+    mod_stream = _yaml2stream(yaml)
+
+    return "{N}:{S}:{V}:{C}:{A}.modulemd.yaml".format(
+        N=mod_stream.get_module_name(), S=mod_stream.get_stream_name(),
+        V=mod_stream.get_version(), C=mod_stream.get_context(),
+        A=mod_stream.get_arch() or "noarch")
```
Could you split NSVCA keys to separate lines for better readability?

dmach commented on modulemd_tools/tests/test_yaml.py:
```diff
@@ -0,0 +1,219 @@
+class TestYaml(unittest.TestCase):
+
+    def test_is_valid(self):
+        assert is_valid(yaml1)
```
Using plain assert is not a good idea.
Asserts can be optimized out if you run python with -OO: ```python3 -OO -c "assert False; print('OK')"```
You should be using self.assert*() methods.

dmach reviewed (changes requested):
Please replace the plain asserts in tests.
The rest is nice to have.

sgallagher commented on modulemd_tools/modulemd_tools/yaml.py:
```diff
@@ -0,0 +1,183 @@
+        A=mod_stream.get_arch() or "noarch")
+
+
+def _yaml2stream(yaml):
+    idx = Modulemd.ModuleIndex.new()
+    idx.update_from_string(yaml, strict=True)
+
+    mod_names = idx.get_module_names()
+    mod = idx.get_module(mod_names[0])
+    mod_stream = mod.get_all_streams()[0]
+    return mod_stream
```
This function should throw an exception if the YAML it was passed results in an index containing more than one module or stream.

FrostyX commented:
Thank you very much for the review @dmach and @sgallagher!

I've addressed all the changes that you suggested and also implemented some new functionality that I needed. The commit history was a total mess, so I squashed everything to the original commit.

> In general, these are indeed useful helper functions. I think they'd be better included in libmodulemd itself, though.
>
>They can be incorporated into 
> https://github.com/fedora-modularity/libmodulemd/blob/main/bindings/python/gi/overrides/Modulemd.py
>
> This would avoid having to pull in a new dependency.

Personally, I don't mind having a new dependency, but I like having this in `libmodulemd` for other reasons. I just felt like it doesn't fit well with the `libmodulemd` concept. I was also afraid that having this in `libmodulemd` would require writing this code in C, so I started this as a separate package. But yeah, I would like to see this in `libmodulemd`.

At this point, I am happy with the functionality this PR provides and I have successfully used these functions in Copr code. Once the Copr PR is ready, I will link it just for the reference. If you would like to move this code into `libmodulemd` itself, please let's re-iterate the review here so we can raise it to a respectable quality and then let's figure out where exactly to move it.

FrostyX commented:
> Once the Copr PR is ready, I will link it just for the reference. 

https://pagure.io/copr/copr/pull-request/1490

sgallagher commented on modulemd_tools/modulemd_tools/yaml.py:
```diff
@@ -0,0 +1,317 @@
+        A=mod_stream.get_arch() or "noarch")
+
+
+def _yaml2stream(mod_yaml):
+    if not mod_yaml:
+        raise ValueError("Provided YAML did not begin with a module document")
+
+    idx = Modulemd.ModuleIndex.new()
+    idx.update_from_string(mod_yaml, strict=True)
+
+    mod_names = idx.get_module_names()
+    if len(mod_names) > 1:
+        raise ValueError("Provided YAML contains multiple modules {0}. "
+                           "It is ambiguous which one to use."
+                           .format(mod_names))
+
+    mod_name = idx.get_module_names()[0]
+    mod = idx.get_module(mod_name)
+    streams = mod.get_all_streams()
+    if len(streams) > 1:
+        raise ValueError("Provided YAML contains multiple streams {0} of "
+                           "a module '{1}'. It is ambiguous which one to use."
+                           .format([x.get_stream_name() for x in streams],
+                                   mod_name))
+    return streams[0]
```
This entire function is unnecessary. You can use
```
Modulemd.ModuleStream.read_string(mod_yaml, strict=True|False)
```

This returns a `Modulemd.ModuleStream` object rather than a `Modulemd.ModuleIndex` and throws an exception if the document contains anything other than a single Stream subdocument.

sgallagher commented on modulemd_tools/modulemd_tools/yaml.py:
```diff
@@ -0,0 +1,317 @@
+        raise ValueError("Provided YAML did not begin with a module document")
+
+    idx = Modulemd.ModuleIndex.new()
+    idx.update_from_string(mod_yaml, strict=True)
```
Be careful about hard-coding `strict=True`. We usually only want to be strict in the tools that are creating repodata or building modules. We don't want to be strict in the tools that *consume* the metadata, because we can have problems when the repodata is created by a newer version of libmodulemd than the one being used by the client (e.g. DNF).

sgallagher commented on modulemd_tools/modulemd_tools/yaml.py:
```diff
@@ -285,27 +285,10 @@ def _generate_filename(mod_yaml):
-                                   mod_name))
-    return streams[0]
+    try:
+        return Modulemd.ModuleStreamV2.read_string(mod_yaml, True, None, None)
```
```suggestion
        return Modulemd.ModuleStream.read_string(mod_yaml, True, None, None)
```

This way, it will work regardless of the metadata version of the stream you are reading in. If you want to ensure that it MUST be v2, do:
```python3
stream = Modulemd.ModuleStream.read_string(mod_yaml, True, None, None)
stream.upgrade(Modulemd.ModuleStreamVersionEnum.TWO)
```

This will raise an exception (`Modulemd.Error.UPGRADE`) if the stream has a higher version than the one you are looking for. If it's version `TWO` or lower, it will automatically be upgraded to v2 and can be worked with normally.

sgallagher commented on modulemd_tools/modulemd_tools/yaml.py:
```diff
@@ -0,0 +1,300 @@
+    Determine whether the `mod_yaml` string is a valid modulemd YAML definition
+    """
+    idx = Modulemd.ModuleIndex.new()
+    ret, _ = idx.update_from_string(mod_yaml, strict=True)
```
This may throw an exception if the YAML itself is invalid and unparseable. (For example, it starts a map or array but never terminates it).

sgallagher commented on modulemd_tools/modulemd_tools/yaml.py:
```diff
@@ -0,0 +1,300 @@
+    if parsed["version"] > version:
+        raise ValueError("Cannot downgrade modulemd version")
+
+    mod_stream = Modulemd.ModuleStream.new(parsed["version"])
```
This is completely useless, as you throw it away on the next line.

sgallagher commented on modulemd_tools/modulemd_tools/yaml.py:
```diff
@@ -0,0 +1,300 @@
+    return idx2.dump_to_string()
+
+
+def upgrade(mod_yaml, version):
+    """
+    Upgrade the input module YAML from an older version to a newer one.
+    Downgrades aren't supported even in case where it would be possible.
+    """
+    parsed = yaml.safe_load(mod_yaml or "")
+    if not parsed or not "version" in parsed:
+        raise ValueError("Missing modulemd version")
+
+    supported = [1, 2]
+    if parsed["version"] not in supported or version not in supported:
+        raise ValueError("Unexpected modulemd version")
+
+    if parsed["version"] > version:
+        raise ValueError("Cannot downgrade modulemd version")
+
+    mod_stream = Modulemd.ModuleStream.new(parsed["version"])
+    mod_stream = mod_stream.read_string(mod_yaml, True,
+                                        parsed["data"].get("name", ""),
+                                        parsed["data"].get("stream", ""))
+    mod_stream_upgraded = mod_stream.upgrade(version)
+    return _stream2yaml(mod_stream_upgraded)
```
This entire function is unnecessary, since everything it is doing is already done by the `ModuleStream.upgrade()` call.

sgallagher commented on modulemd_tools/modulemd_tools/yaml.py:
```diff
@@ -0,0 +1,300 @@
+    Same as `is_valid` but raises exception if the YAML is not valid.
+    """
+    idx = Modulemd.ModuleIndex.new()
+    ret, failures = idx.update_from_string(mod_yaml, strict=True)
```
See above

FrostyX commented:
Thank you for the review @sgallagher,
I've addressed all the things that you pointed out.

FrostyX commented:
I was reflecting on this several times and I decided to lower my ambitions regarding this python library. It was created to 1) share code between the tools provided by this project 2) to be used by other tools outside of this repository. I am going to omit the second part.

Currently, we are working on test coverage for this project and it would make the task much easier if we could re-use functions from this library (since they are well-tested and IMHO immensely useful), throw away a lot of duplicities, and then test the rest. I decided to merge the implementation as is and add a warning discouraging the general public from using it. If we decide to pick pieces of the library and move them to the `python3-lilbmodulemd` in the future, it won't be any problem from within this project, as it would mean just changing imports.

I hope it makes sense.

Thank you guys for the review, we fixed a lot of things here, and I am happy they won't kick me later.
